### ISSUE TITLE
SDL2 install on Ubuntu 20.04, menu not working, including multiwindows, tiles, and font sizes

### ISSUE THREAD

**Issue by jason-mehmel** [angband/angband#4762]
After getting some help to get the SDL2 frontend working, I can now run angband 4.2.2!

(We figured it out here: https://github.com/angband/angband/issues/4745)

I was hoping to be able to:

a)be able to change font sizes and choose tiles.
b)use the multiwindows
c)run it windowed.

But in SDL2, currently, it only opens in fullscreen, with one main screen in the centre and the repeating angband logo around it.

It SEEMS like the tiles and other fonts aren't installed, maybe? Do I need to download those separately? But also, the menu doesn't seem to be working.

I'll detail what I can do in the menus. (It isn't much.)

The menu allows me to click on fullscreen, but to no effect; it doesn't toggle to a non-fullscreen window.

I can click on Angband>Font>Name>Size, but all options are greyed out, other than 10x20.fon

I can click on Angband>Tiles>Set and Size, but the options beyond are greyed out.

I can click on Angband>Borders but no other options come up and nothing changes.

Going to Angband>Top is greyed out.

Set Keypad Modifier is greyed out.

Windows>Window-0 is available, but the other ones are greyed out.

About is available, but nothing happens on clicking.

The top A 1 2 3 4 5 6 7 options are all greyed out other than 'A'.

Quit also does nothing.

**Comment by NickMcConnell** [comment]
Greyed out mostly means that's not the active choice at the moment.  Try mousing over Angband>Tiles>Set and then selecting a tile set; other things should work similarly.

**Comment by jason-mehmel** [comment]
That's the thing; none of the tiles are an active choice, ditto for the other subwindow options.

Honestly, none of the menu actually *works,* I can go through the menu items but none of my selections (including **Fullscreen** and **Quit**) do anything.

**Comment by jason-mehmel** [comment]
For reference, two screenshots showing the greyed-out options. My assumption is all the tiles and fonts that are in the folders should be available.

Is there a config file I need to adapt somewhere to get those available? And the sub windows?

![angband error 2](https://user-images.githubusercontent.com/3622823/116154553-0692d480-a6a6-11eb-84ed-f04b30884878.png)
![angband error](https://user-images.githubusercontent.com/3622823/116154556-072b6b00-a6a6-11eb-8363-78b3eaa7ef24.png)

**Comment by NickMcConnell** [comment]
OK, for me it looks exactly like your first screenshot, but I can go to one of the tilesets and **even though it's greyed out** select it, and it works.

**Comment by jason-mehmel** [comment]
Ah! That makes sense for the confusion. I **can** click on the options, but nothing changes.

Which is what's making me wonder if there's a config file I'm missing, or still a glitch connected to my previous issue in the process of compiling and locally installing?

**Comment by backwardsEric** [comment]
There is a configuration file that stores the settings between runs, but it seems unlikely that its presence/absence/corruption would cause this.  If you wanted to start with a clean slate for that, you could move ~/.angband/Angband/sdl2init.txt to somewhere else (or just remove it if you're not interested in retaining any settings), and then run Angband.

As for something being set incorrectly during the build process, you might attach mk/buildsys.mk and mk/extra.mk here (if it complains about them being an unsupported file type, you can first tar and compress and attach that, i.e.

`tar -cf - mk/buildsys.mk mk/extra.mk | gzip -c - > ~/mk.tgz`

); that would let us see what was set for compiling.  You could then run

`make clean
rm mk/buildsys.mk mk/extra.mk
./configure --with-no-install --enable-sdl2
make`

which should get you a clean build.

**Comment by jason-mehmel** [comment]
Okay! I've got the MK file attached as a ZIP. Removing the sdl2init.txt didn't fix the problem, and neither did the process of removing the **buildsys** and **extra** MK files.

[mk.zip](https://github.com/angband/angband/files/6389282/mk.zip)

So, if it's not a matter of the files being missing (I can find them in the file structure) and it's not a matter of the build setting things up wrong in a config sense, what other reason could there be for no response from the window options?

And a possibly dumb question: Is the SDL2 version the only version that I can use to view and use tiles?

**Comment by backwardsEric** [comment]
One possibility would be some sort of incompatibility between the front end and the version of SDL2 that Ubuntu 20.04 has (it looks like that is 2.0.10; the Debian Buster system where it worked for me has 2.0.9; I don't know the configuration Nick used).  If there are multiple versions of a library installed (or a botched upgrade that lefts bits of one version and some of another) then there could be the possibility that the header files used when compiling don't match up with the library that is linked or pulled in at runtime - though that shouldn't be the case here as you had newly installed the SDL2 packages.

There is a front end based on SDL1.2 that allows the use of tiles,  On Debian Buster, the packages needed for that are libsdl-image1.2-dev, libsdl-mixer1.2-dev, libsdl-ttf2.0-dev, and libsdl1.2-dev.  It looks like packages with the same names are available for Ubuntu 20.04.  For the arguments to use with configure,
you'd use

'''./configure --with-no-install --enable-sdl'''

.  Then, after running make, you'd run

'''src/angband -msdl'''

to bring up that front end.

The Windows front end also supports tiles.  Assuming your system has a x86_64 CPU, you could build the windows executable and run it under wine.  The mingw cross-compiler and wine are available as packages so it is easy enough, but that's more likely to run into problems that haven't already been seen and fixed.

**Comment by jason-mehmel** [comment]
Thank you once again @backwardsEric , that worked getting the SDL version working. (Not the SDL2, to avoid confusion here.) I also ran `make clean` since I haven't downloaded the files since the updates from the last problem I reported!

You've got an interesting theory about SDL2 not working on Ubuntu 20.04 due to a slight incompatibility. Happy to help test that further if need be!

Now, for my next unrelated question: how do I control the output of the subwindows in this SDL frontend? Is that a config file?

**Comment by backwardsEric** [comment]
For configuring the output of the subwindows, make the subwindows you want visible:
1. In the top menu bar, click on the yellow label to the right of "Term:" and select the subwindow from the menu that appears.
2.  Then click on the yellow label to the right of "Visible:" in the top menu bar to set the visibility of the subwindow that's currently selected from the "Term:" menu.

Then with a character loaded in the game, press '=' to bring up the options menu and then press 'w' while in that menu to edit the table of which subwindows are flagged for displaying particular content.  Back in the main options menu, you can save those subwindow settings to a preference file by pressing 's'.  The help at https://angband.readthedocs.io/en/latest/customize.html#user-pref-files has more information about using preference files.

**Comment by jason-mehmel** [comment]
Thank you @backwardsEric !

I think this issue is, if not closed, at least finished for me. (I think the greater question about Ubuntu 20.04 might still be something to solve?)

**Comment by backwardsEric** [comment]
Yes it is.  I tried installing the latest version of SDL2 (2.0.14) in my own directories on Debian Buster and building Angband with that and did not encounter the trouble you had with the version built against the Ubuntu package - clicking on the menu entries simply worked.  The next experiment would be with Ubuntu 20.04.  Could you let us know what desktop environment or window manager you were using when the SDL2 front end did not work?  Perhaps that might be a factor.

**Comment by backwardsEric** [comment]
I couldn't reproduce the issues with the SDL2 menus not working.  I set up Ubuntu 20.04 as a virtual machine in Parallels 16.5.0 (using the built-in option in Parallels to do so), used 'sudo apt-get install' to install the needed packages (autoconf, automake, make, libsdl2-dev, libsdl2-image-dev, libsdl2-ttf-dev, libsdl2-mixer-dev), and then built and ran the current Angband code (./autogen.sh; ./configure --with-no-install --enable-sdl2; src/angband -msdl2).

**Comment by jason-mehmel** [comment]
Here's one thing I've been meaning to follow up on. Until recently, I was using the Wayland video manager. I've switched back to Xorg, and I'm wondering if that might have contributed to the issue. I've got this in my to-do list but it's a little lower down than some other hot button issues for me.

I'll check two things: trying SDL2 in xorg, and then going back to Wayland and trying it.

**Comment by fumsoft** [comment]
Interesting. I'll be keeping an eye on this one.
That reminds me of something. OP, are you using a mouse, a touchpad or what?

> So, if it's not a matter of the files being missing (I can find them in the file structure) and it's not a matter of the build setting things up wrong in a config sense, what other reason could there be for no response from the window options?

Bugs in main-sdl2.c, bugs in SDL2, bugs in Wayland...

**Comment by kirtr** [comment]
I encountered this problem as well. I couldn't reproduce the bug by using live usb installs and compiling and testing on various versions. The live versions apparently default to xorg. It wasn't until I saw this bug that I tried reverting to Xorg from Wayland. The problem vanished. Went back to Wayland; problem back. Back to Xorg, problem gone.

SDL2 and Wayland are definitely the culprits. I'd add with Wayland to the name of the bug.

Ubuntu version: 21.04 Hirsute Hippo
libwayland-server0:amd64  1.18.0-2~exp1.1      wayland compositor infrastructure - server library
xwayland                              2:21.1.1-0ubuntu1  Xwayland X server
libsdl2-2.0-0:amd64             2.0.14+dfsg2-3       Simple DirectMedia Layer

**Comment by fumsoft** [comment]
BTW, I think SDL2 uses xwayland by default? (as of 2.0.16)
Do you "export SDL_VIDEODRIVER=wayland" prior to launching Angband?
Also, does mouse/touchpad work for actually playing the game - running around and stuff?
(if it does, the most likely thing is probably xwayland forgetting to set window focus... see main-sdl2.c:3133)

**Comment by fumsoft** [comment]
> (if it does, the most likely thing is probably xwayland forgetting to set window focus... see main-sdl2.c:3133)

Or actually it's not that - I forgot that (according to the OP) mouse motion events do work.

**Comment by ShimmerFairy** [comment]
I'm having this issue as well, with Angband 4.2.3, on X11. None of the menus work at all. I can open the menu and submenus, but clicking on any menu items does nothing but get the open menu stuck on screen. I can't undo fullscreen mode, I can't resize or move any windows, and the only way I can exit the program is by alt-tabbing and killing the game.

**Comment by backwardsEric** [comment]
From a report by Michaelyx, https://angband.live/forums/forum/angband/vanilla/249611-experiences-with-new-angband-4-2-5-and-questions-about-identifying-and-variants?p=252048#post252048 , his issues with nonfunctional menus in the SDL2 front end on Ubuntu 24.04 went away when he used the latest post-4.2.5 build.  There's been many changes to the SDL2 front end since 4.2.5 but the one that's most likely to affect this problem is this one, https://github.com/angband/angband/commit/91dc81260ff82bff02ada27caa96b7d68e9b04ba , which is present in the post-4.2.5 versions since 4.2.5-77-g91dc81260.  In 4.2.5, the menu handling in the SDL2 front end would process the first mouse motion event in the queue and then call SDL_FlushEvent(SDL_MOUSEMOTION) to flush any others that were pending.

**Comment by NickMcConnell** [comment]
4.2.6 would probably be a good idea before too long, I'll put that on my todo list...

**Comment by backwardsEric** [comment]
Two things that I thought might be incorporated into 4.2.6 but which are not close to done:


1. Implement a font browser for the SDL2 front end to select fonts outside the lib/fonts directory.
2. Rebalance how difficult it is to use wands, staves, rods, dragon armors, and artifact activations for the change to the device failure rate made in 4.2.3.

**Comment by NickMcConnell** [comment]
Those sound sensible - are there issues for them?  Also I have been letting triage - and actual 4.2 - issues pile up, and I'm intending to maintain my focus on Beleriand for a bit, so when I say "before too long" I probably mean months rather than weeks 😁

**Comment by NickMcConnell** [comment]
Brilliant, thanks

**Comment by FlorentFlament** [comment]
Hi, I have the same issue (SDL2 frontend not working) on the version 4.2.5 of Angband packaged in Debian 13 (Trixie). On version 4.2.5-512-g6b168aff3 the issue is resolved.

**Comment by kirtr** [comment]
@NickMcConnell, how much work is it to release a 4.2.6? I'd still love to get Angband 4.x working properly in Debian / Ubuntu for the first time! A new release would greatly ease that process for the upstream guys I have to push on.

**Comment by NickMcConnell** [comment]
> [@NickMcConnell](https://github.com/NickMcConnell), how much work is it to release a 4.2.6? I'd still love to get Angband 4.x working properly in Debian / Ubuntu for the first time! A new release would greatly ease that process for the upstream guys I have to push on.

See the Jan 30 and 31 comments from @backwardsEric above for stuff he still wants to put in.  I guess it's a balance between including those things and just getting 4.2 into the Debian-based distros.

EDIT: In fact the second of those was done in #6325, so it's just the SDL2 font selection in #3357 remaining.

**Comment by backwardsEric** [comment]
Yes, and the SDL2 font selection is getting closer.  Probably a week to two to have it ready for a pull request, and then some time for others to test it.  If we want the 4.2.6 release before the end of the year, the SDL2 font selection could well be skipped.

**Comment by NickMcConnell** [comment]
I'm inclined to say wait if it's that close.

**Comment by kirtr** [comment]
That is fantastic progress @backwardsEric! I think we should release as soon as possible. Here's why:
- the next Ubuntu LTS (long term support) release is being built now
- other major distributions such as Linux Mint and Pop! OS are based on Ubuntu LTS
- so, if we don't get a working Angband into Ubuntu very soon, we'll have to wait until **April 2028** for that to happen
- it needs to first get put into Debian since Ubuntu pulls packages from Debian
- we need enough time to get the maintainers, like Magnate, to be able to pull it in and debug it
- the last release to "just work" without reading github issues and massaging it was 3.5
- it has just been way too long without a working Angband in some of the most popular Linux distros
- having a working Angband in some of the most popular distributions would likely bring new people and energy into Angband

Based on this information, what do you guys think about a release sooner rather than later? I'm concerned that our timelines are already very tight.

**Comment by NickMcConnell** [comment]
OK, I'm also fine with release ASAP 🙂

**Comment by FlorentFlament** [comment]
First, thank you for all the passion put into this awesome game, and to make it available on major distributions.  I'm a new Angband player (Lvl 21 Half-Orc Warrior), and a believer that having a recent working version packaged in major distributions is the last step for great software.  And I believe that Angband deserves this, so that it be easily available to the greater Free Software community.  Thanks again ! ; )

**Comment by backwardsEric** [comment]
As I see it, these pending pull requests, #6410 , #6411 , #6412 , #6413 , and #6415 , should be applied prior to the release.  Sphinx is issuing some diagnostics about docs/hacking/compiling.rst so a pull request to fix those and some block quoting of commands in that file should also be applied.  I have a draft of that, but it will need to be adjusted for #6410 .  #6416 is the first of a handful of pull requests to check off things in the release checklist.  I'm working on draft of updates to changes.txt and the list of contributors in docs/thanks.rst for the release.  I'll try to have that posted today.

**Comment by NickMcConnell** [comment]
Wonderful, I've pulled those in - I hope that wasn't premature.

**Comment by backwardsEric** [comment]
The changes to the compiling documentation is at https://github.com/angband/angband/pull/6418 .  The draft of the update for changes.txt is at https://github.com/angband/angband/pull/6421 .

**Comment by NickMcConnell** [comment]
I've pulled in #6418 and #6419.  I' not sure if #6420 is final, and I'm guessing you may want to do #6421 right before release, so I've left those two for now

**Comment by kirtr** [comment]
Thank you both for all of the hard work! Looking forward to trying to get it working, out of the box, on Debian, Ubuntu, etc with SDL2 and Wayland. Hopefully we can close this bug when released and that will be my trigger to see what I can help with on the Debian based distributions.

**Comment by NickMcConnell** [comment]
4.2.6 has been released now.  I think this can now be closed, but I'll wait for confirmation from @backwardsEric who is now the official maintainer of Angband 😄

**Comment by backwardsEric** [comment]
Yes, 4.2.6 is up at https://github.com/angband/angband/releases/tag/4.2.6 .  It includes the change to the desktop file for SDL2 (the problem in the second part of this thread) and changes how the SDL2 front end handles consecutive queued mouse motion events (evidently the cause for the unresponsive menus mentioned in the first part of this thread and in Debian's bug list as https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1121870 ).

It also fixes two other bugs in Debian's list:  damage notifications for spells ( https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=780770 ) and incorrect spoiler output ( https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=781748 ).

**Comment by NickMcConnell** [comment]
All right, closing as requested by @kirtr
